src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -265,46 +293,130 @@ private BoundExpression MakePatternIndexOffsetExpression(BoundExpression unlower
             MethodSymbol sliceMethod,
             BoundExpression rangeArg)
         {
-            // Lowered code:
+            Debug.Assert(TypeSymbol.Equals(
+                rangeArg.Type,
+                _compilation.GetWellKnownType(WellKnownType.System_Range),
+                TypeCompareKind.ConsiderEverything));
+
+            // Lowered code without optimizations:
             // var receiver = receiverExpr;
             // int length = receiver.length;
             // Range range = argumentExpr;
             // int start = range.Start.GetOffset(length)
-            // int end = range.End.GetOffset(length)
-            // receiver.Slice(start, end - start)
+            // int rangeSize = range.End.GetOffset(length) - start
+            // receiver.Slice(start, rangeSize)
```

Comment:
> receiver.Slice(start, rangeSize) [](start = 14, length = 33)

Is there a design doc for ranges, where we can document the API patterns recognized by the compiler?

Reply:
https://github.com/dotnet/csharplang/blob/8152182f0a477cb3082e625b607262cc459a17f3/proposals/index-range-changes.md